### PR TITLE
Fractions

### PR DESCRIPTION
**Pull request by aliang8**
Fixed `COMMON_DENOMINATORS` to work with non integer denominators.

### CODE REVIEW

File: lib/rule-list.js [hunk 1]
```diff
@@ -155,17 +155,95 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+    (node) => {
+        if (query.isAdd(node)) {
+            const terms = node.args
+            // checks if at least one of the terms is a fraction
+            const hasFraction = terms.some(term => query.isDiv(term))
+
+            let sameDenom
+            // rule cannot apply if all of the terms are fractions
+            // and they all have the same denominator
+            if (terms.every(term => query.isDiv(term))) {
```
Comment (k4b7) [inline lib/rule-list.js]: We should avoid requiring every single term b/c it prevents handling of common denominators of fractions surrounded by non-fraction terms, e.g.
```
1 + 2/x + 3/y + 4
```

Comment (k4b7) [reply lib/rule-list.js]: I think you should be able to run `matchNode` on the current node with the `#a_0 / #b_0 + ...` as the pattern to get the start/end indexes of the first group of 2 or more consecutive fractions.  You'll have the pattern to a matching function before you can use it with `matchNode`.

File: lib/rule-list.js [hunk 2]
```diff
@@ -155,17 +155,95 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+            denoms = denoms.map(denom => query.getValue(denom))
+            const LCM = lcm(...denoms)
+            newDenoms = denoms.map(
+                denom => build.mul(build.number(denom),
+                                   build.number(LCM / denom)))
+            newNumerators = nums.map(
+                (num, i) => build.mul(num,
+                                      build.number(LCM / denoms[i])))
+        } else {
+            newDenoms = build.mul(...denoms)
```
Comment (k4b7) [inline lib/rule-list.js]: This isn't ideal, if all of the denominators are `x` then we'll make this overly complicated.

File: lib/rule-list.js [hunk 3]
```diff
@@ -155,17 +155,95 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+            }
+
+            return hasFraction && !sameDenom ? {node} : null
+        }
+    },
+
+    (node) => {
+        const terms = node.args
+        // an array storing the index where fraction is negative
+        const negatives = terms.map(term => query.isNeg(term))
```
Comment (k4b7) [inline lib/rule-list.js]: It would be could if we could extract this logic for reuse.  Don't worry about it for this review, but something to think about.

File: lib/__test__/rule-list.test.js [hunk 4]
```diff
@@ -181,12 +188,14 @@ describe('rules', () => {
 
     suite('common denominators', rules.COMMON_DENOMINATOR, [
         ['2/6 + 1/4', '(2 * 2) / (6 * 2) + (1 * 3) / (4 * 3)'],
+        ['2/(3 + x) + 3/(2 + x)', '(2 * (2 + x)) / ((3 + x) * (2 + x)) + (3 * (3 + x)) / ((3 + x) * (2 + x))'],
```
Comment (k4b7) [inline lib/__test__/rule-list.test.js]: There should be some tests with:
- a mix of numeric denominators and variable denominators
- repeated variable denominators
- repeated numeric denominators
- non-fractional terms in the sum

**Review by k4b7 — CHANGES_REQUESTED** [review]

File: lib/__test__/rule-list.test.js (hunk 4 above)
Comment (aliang8) [reply lib/__test__/rule-list.test.js]: COMMON_DENOMINATORS won't match when the denominator is the same throughout regardless of whether they are variable or numeric. I'm not entirely sure how we want to handle non-fractional terms in the sum yet.

File: lib/rule-list.js (hunk 2 above)
Comment (aliang8) [reply lib/rule-list.js]: When will all the denominators be `x`? I don't think COMMON_DENOMINATORS will match that.

File: lib/rule-list.js (hunk 3 above)
Comment (aliang8) [reply lib/rule-list.js]: What logic are you referring to? The negatives array logic or the rule itself?

File: lib/rule-list.js (hunk 1 above)
Comment (aliang8) [reply lib/rule-list.js]: I see, I didn't consider using `matchNode` inside of a `defineRule` function. Wouldn't that pattern not take into consideration the case of non-fraction terms (`1 + 2/x + 3/y + 4`)?

File: lib/__test__/rule-list.test.js (hunk 4 above)
Comment (k4b7) [reply lib/__test__/rule-list.test.js]: There are a couple of options:
- don't include non-fractions when matching
- include non-fractions, but convert them to fractions before computing the common denominator.

You should ask Gen for their input on this.

What happens if some of the fractions have the same denominator while others don't?

Comment (aliang8) [reply lib/__test__/rule-list.test.js]: Yup good points. I had a test case that was commented out addressing this issue. I will ask for Gen's input. 

> What happens if some of the fractions have the same denominator while others don't?

It works fine in this scenario. I've added a test case for this. The only time when this rule doesn't match is when all the terms are fractions and they all have the same denominator.

Comment (aliang8) [reply lib/__test__/rule-list.test.js]: I added a fix for it.

File: lib/rule-list.js (hunk 3 above)
Comment (k4b7) [reply lib/rule-list.js]: Computing the `negatives` array from the terms.

Comment (aliang8) [reply lib/rule-list.js]: Fixed!

File: lib/rule-list.js [hunk 5]
```diff
@@ -155,17 +155,181 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+                          ? decimal_to_fraction(term.args[0])
+                          : isDecimal(term)
+                          ? decimal_to_fraction(term)
+                          : isPolynomialTerm(term) && query.isNeg(term)
+                          ? build.div(term.args[0], build.number(1))
+                          : isPolynomialTerm(term)
+                          ? build.div(term, build.number(1))
+                          : query.isNeg(term)
+                          ? term.args[0]
+                          : term)
```
Comment (k4b7) [inline lib/rule-list.js]: This is super confusing.  Can you rewrite this use `if-else` statements?

Comment (aliang8) [reply lib/rule-list.js]: Yeah xD I realized. Sure.

File: lib/rule-list.js [hunk 6]
```diff
@@ -155,17 +155,181 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+    (node) => {
+        if (query.isAdd(node)) {
+            const terms = node.args
+            
+            // checks if at least one of the terms is a fraction
+            const hasFraction = terms.some(term => isFraction(term))
+ 
+            let sameDenom
+            // rule cannot apply if all of the terms are fractions
+            // and they all have the same denominator
```
Comment (k4b7) [inline lib/rule-list.js]: It might be worth moving this comment up to the top of the function so that people don't have to read the whole function to understand the restrictions.

File: lib/rule-list.js [hunk 7]
```diff
@@ -155,17 +155,181 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+            if (terms.every(term => isFraction(term))) {
+                const denom = getDenominator(terms[0])
+
+                sameDenom = terms.every(term =>
+                                        print(getDenominator(term)) == print(denom))
```
Comment (k4b7) [inline lib/rule-list.js]: We might want to encapsulate this into an `equals` function in `utils.js`.

File: lib/rule-list.js [hunk 8]
```diff
@@ -155,17 +155,187 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+            }
+
+            return hasFraction && !sameDenom ? {node} : null
+        }
+    },
+
+    (node) => {
+        node = canApplyRule(CONVERT_TO_FRACTION, node)
+            ? applyRule(CONVERT_TO_FRACTION, node)
+            : node
```
Comment (k4b7) [inline lib/rule-list.js]: We convert non-fraction terms to fractions here, but in the match function part of this rule we return false if any of the terms are non-fractions.  Am I missing something?

File: lib/rule-list.js [hunk 9]
```diff
@@ -155,17 +155,187 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+
+export const getNumerator = (node) => query.isNeg(node)
+    ? node.args[0].args[0]
+    : node.args[0]
+
+export const getDenominator = (node) => query.isNeg(node)
+    ? node.args[0].args[1]
+    : node.args[1]
+
+export const isDecimal = (node) => query.isNumber(node) && query.getValue(node) % 1 != 0
```
Comment (k4b7) [inline lib/rule-list.js:181]: Can you add a TODO to move these helper functions to `math-nodes`?

File: lib/rule-list.js [hunk 10]
```diff
@@ -155,17 +155,187 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+            const LCM = lcm(...denoms)
+
+            // remove multiplication by 1
+            newDenoms = denoms.map(
+                denom => denom == 1 
+                    ? build.number(LCM / denom)
+                    : LCM / denom == 1
+                    ? build.number(denom)
+                    : build.mul(build.number(denom),
+                                build.number(LCM / denom)))
```
Comment (k4b7) [inline lib/rule-list.js]: These chained ternaries are really hard to understand.  Can you rewrite this using `if-else` statements?

File: lib/rule-list.js [hunk 11]
```diff
@@ -155,17 +155,187 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+            newNumerators = nums.map(
+                (num, i) => LCM / denoms[i] == 1
+                    ? num
+                    : build.mul(num,
+                                build.number(LCM / denoms[i])))
+        } else {
+            newDenoms = build.mul(...denoms)
```
Comment (k4b7) [inline lib/rule-list.js]: What happens for `2 / x + 3 / x + 4 / y`?  Can you add a tests case like this?  I would expect the result to be `2y / xy + 3y / xy + 4x / xy`.

File: lib/rule-list.js [hunk 12]
```diff
@@ -155,17 +155,187 @@ export const SIMPLIFY_FRACTION = defineRuleString(
...
+            newNumerators = nums.map(
+                (num, i) => build.mul(num, ...denoms.filter(e => e != denoms[i])))
```
Comment (k4b7) [inline lib/rule-list.js]: What about situations where there are multiple factors in the denominator to begin with?  I think there's an implicit assumption that each fraction as a single factor denominator.  Can you add a test case for `2 / (x * x) + 3 / (x * y) + 4 / (y * y)`?

File: lib/rule-list.js (hunk 8 above)
Comment (aliang8) [reply lib/rule-list.js]: The first part is correct. The match function of `COMMON_DENOMINATORS` returns false when all the terms are fractions and they all have the same denoms or none of the terms are fractions.

File: lib/rule-list.js (hunk 11 above)
Comment (aliang8) [reply lib/rule-list.js]: Ah, good catch. Will fix.

File: lib/rule-list.js (hunk 12 above)
Comment (aliang8) [reply lib/rule-list.js]: I see. In `ChangeTypes`, `MULTIPLYING_POLYNOMIALS`comes before `COMMON_DENOMINATORS`, so we won't see this exact case. We would see, however, `2/ x^2 + 3/ x^1y^1 + 4 / y^2` which we would want to simplify to ` (2 * y)/ x^2y^1 + (3 * x) / x^2y^1 + (4 * x) / x^2y^1`.

Comment (aliang8) [reply lib/rule-list.js]: A similar test case that we should also consider is:
`2 / ((x^2 + 1)^2 * (x + 1)^1) + 3 / ((x^2 + 1)^2 * x)`

File: lib/rule-list.js (hunk 8 above)
Comment (k4b7) [reply lib/rule-list.js]: I see, it has to have at least one fraction... I obviously need to get more sleep.  :P

**Comment by k4b7** [comment]
This looks good except for the failing test.  Any idea what's going on there?  It looks like the rule returns an empty string.

**Comment by aliang8** [comment]
This PR is not ready to merge. I am still working on taking care of more edge cases.

**Comment by k4b7** [comment]
@aliang8 is this ready for review now?
